research/2022-11-02-building-privacy-protecting-infrastructure.md

```diff
@@ -0,0 +1,489 @@
+they are not using the protocol directly themselves as part of normal operation,
+or if they aren't socially inclined to provide a certain service. This depends a
+lot on how an individual platform decides to use the network.
+
+![Waku Network](/img/building_private_infra_network.png)
+
+### Dealing with network spam and RLN Relay
+
+Since the p2p relay network is open to anyone, there is a problem with spam. If
+we look at existing solutions for dealing with spam in traditional messaging
+systems, a lot of entities like Google, Facebook, Twitter, Telegram, Discord use
+phone number verification. While this is largely sybil-resistant, it is
+centralized and not private at all.
+
+Historically, Whisper used PoW which isn't good for heterogenerous networks.
+Peer scoring is open to sybil attacks and doesn't directly address spam
+protection in an anonymous p2p network.
+
+The key idea here is to use RLN for private economic spam protection using
+zkSNARKs.
+
+I'm not going to go into too much detail of RLN here. If you are interested, I
+gave a [talk](https://www.youtube.com/watch?v=g41nHQ0mLoA) in Amsterdam at
+Devconnect about this. We also have some write-ups on
+[here](https://vac.dev/rln-relay) Sanaz who has been pushing a lot of this from
+our side. There's also another talk at Devcon by Taylor going into RLN in more
+detail.
+
+I'll briefly go over what it is, the interface and circuit and then talk about
+how it is used in Waku.
+
+### RLN - Overview and Flow
+
+RLN stands for Rate Limiting Nullifier. It is an anonyomous rate limiting
+mechanism based on zkSNARKs. By rate limiting we mean you can only send N
+messages in a given period. By anonymity we mean that you can't link message to
+a publisher. We can think of it as a voting booth, where you are only allowed to
+vote once every election.
+
+![Voting Booth](/img/building_private_infra_vote.png)
+
+It can be used for spam protection in p2p messaging systems, and also rate
+limiting in general, such as for a decentralized captcha.
+
+There are three parts to it. You register somewhere, then you can signal and
+finally there's a verification/slashing phase. You put some capital at risk,
+either economic or social, and if you double signal you get slashed.
+
+### RLN - Circuit
+
+Here's what the private and public input to the circuit looks like. The identity
+secret is generated locally, and we create an identity commitment that is
+inserted into a merkle tree. We then use merkle proofs to prove membership.
+Registered member can only signal once for a given epoch or external nullifier,
+for example every ten seconds in Unix time. RLN identifer is for a specific RLN
+app.
+
+We also see what the circuit output looks like. This is calculated locally. `y`
+is a share of the secret equation, and the (internal) nullifier acts as a unique
+fingerprint for a given app/user/epoch combination. How do we calculate `y` and
+the internal nullifier?
+
+```
```

Comment:
Not sure, but something like this may work (or any language close to circom in terms of syntax)
```suggestion
```circom
```